Fix Version#compareTo() implementation

ssoloff opened the pull request:
Based on a discussion in #2483.  Per the `Comparable#compareTo()` method contract defined in the JDK, this method must:

1. throw an NPE when `other` is `null`
1. return a negative integer when `this` is less than `other`
1. return a positive integer when `this` is greater than `other`

The `Version#compareTo()` implementation did not do (1), and (2) and (3) were reversed.  This PR fixes these issues.  Other changes in this PR include:

* Throw an NPE from `Version#isGreaterThan()`, `Version#isLessThan()`, and `Version#isCompatible()` if `other` is `null`.
* Replace the custom `Version#equals()` tests with Equals Verifier.

#### Testing

New unit tests were added to ensure all code paths through `Version#compareTo()` are covered.

With respect to the changes that cause an NPE to be thrown when `other` is `null`, I verified that all  callers to the affected methods pass a non-`null` value.  Note that there may be some indirect callers to `Version#compareTo()` (e.g. in JDK code) that I couldn't examine, but, again based on the `Comparable#compareTo()` method contract, these callers should never be passing `null` to this method, and it should be considered a programming error if they do.

ssoloff commented on src/test/java/games/strategy/engine/framework/startup/mc/GameSelectorModelTest.java:
```diff
@@ -105,8 +101,7 @@ private void verifyTestObjectObserverUpdateSent() {
   private void prepareMockGameDataExpectations() {
-    when(mockGameData.getGameVersion()).thenReturn(mockVersion);
-    when(mockVersion.toString()).thenReturn(fakeGameVersion);
+    when(mockGameData.getGameVersion()).thenReturn(new Version(fakeGameVersion));
```
This change was triggered by making `Version` `final`, which itself was triggered by Equals Verifier.  Making `Version` `final` is a reasonable change since it is effectively a value type.

However, once `Version` is `final`, it can no longer be mocked.  The mocking of `Version` in this test seemed unreasonable anyway, as we were essentially subverting the `Version` class invariants by using an invalid version string.  I simply replaced the test value with a valid, but unlikely, version string.

DanVanAtta replied:
Okay, though may as well store `Version` as a class variable.
I'm not sure about subversion of the class invariant, though it does not seem there is a stronger guarantee of valid data and so completely bogus data in a test case will better exercise the code to demonstrate that we handle this data (and also demonstrates that the code does not have proper invariant checks in place and that our test data *could* look like this, which is for sure a smell. The fix is to add an invariant that then requires this test to be udpated to have stronger data. I generally will use the absolute worst case data I can think of for a test and still have the code worked. It can show the lack of integrity checks in the code)

ssoloff replied:
> Okay, though may as well store `Version` as a class variable.

Not sure I follow here.  The `Version` instance isn't used outside of this method.  Why would we promote it to class level?  Contrast this with the `Integer` instantiated two lines below.  It's basically the same usage.  Should that be extracted to a field, as well?

> I'm not sure about subversion of the class invariant, though it does not seem there is a stronger guarantee of valid data

There is such an invariant in place: the `Version(String)` constructor will raise an exception if you pass `"fakeGameVersion"`.

DanVanAtta replied:
> The Version instance isn't used outside of this method. 

Good point, meanwhile we do use the inner String value elsewhere. I'm okay with leaving as is then.

RoiEXLab commented on src/main/java/games/strategy/util/Version.java:
```diff
@@ -204,7 +211,16 @@ public String toString() {
+   * @param other The version to compare to this version for compatibility.
+   *
+   * @return {@code true} if the specified version is compatible with this version; otherwise {@code false}.
+   */
```
We should perhaps mention, that passing null results in a NPE.
`throws NullPointerException When other is null`

RoiEXLab replied:
(a null value is obviously not compatible, but a programmer could expect it to return false similar to equals)

ssoloff replied:
There was a discussion on this topic a few months ago because I was going OCD on documenting nullity for method parameters and return values.  At that time, @DanVanAtta recommended that we adopt the convention that all parameters and return values are assumed to be non-`null` unless stated otherwise in the Javadocs, and that passing a `null` argument will result in an NPE.

Since that time, we've also introduced the JSR 305 nullity annotations.  Applying `@Nullable` to a method parameter or return type to indicate `null` is an acceptable value seems to be an improvement over explicitly writing it out in the Javadocs.  The lack of such an annotation on `other` should imply that passing `null` for that parameter will result in an NPE.

RoiEXLab commented on src/main/java/games/strategy/util/Version.java:
```diff
@@ -131,42 +135,41 @@ public int hashCode() {
   @Override
-  public int compareTo(final Version o) {
-    return o == null ? -1 : compareTo(o, false);
+  public int compareTo(final Version other) {
```
Are the only calls to this method from within this class?
Or are all other calls comparing with 0?
In the latter case we should probably just call equals instead.

ssoloff replied:
There aren't any direct calls to `compareTo(Version)`.  I thought there may be some indirect calls from the `Comparator` implementations in the lobby client UI, but the "version" columns displayed in the lobby are actually strings, not `Version`s.  I set a breakpoint in `compareTo(Version)` and connected to the lobby; the breakpoint was never hit from the UI code.  That may imply that implementing `Comparable` could be removed.

However, the `isGreaterThan()` and `isLessThan()` methods are still required.  Their presence indicates that `Version` _should_ implement `Comparable`.  Since `Comparable` is the canonical way to determine if one instance is less/greater than another, I'm wondering if we should actually remove `isGreaterThan()` and `isLessThan()` and replace them with calls to `compareTo()`?  The obvious problem with that at this time is the additional overload `isGreaterThan(Version, boolean)`.

RoiEXLab commented on src/main/java/games/strategy/util/Version.java:
```diff
@@ -131,42 +135,41 @@ public int hashCode() {
+    return compareTo(other, false);
   }
 
   private int compareTo(final Version other, final boolean ignoreMicro) {
```
I really think this overload is a code smell.
All calls to this method with ignoreMicro = true can be replaced with a call to isCompatible (and optionally, additionally isGreaterThan/isLessThan if the call does not compare against 0)

ssoloff replied:
I agree that this is a smell.

There's only one call that will invoke this method with `ignoreMicro` equal to `true`, and that's in `GameParser#parseMinimumEngineVersionNumber()`:

```java
private void parseMinimumEngineVersionNumber(final Node minimumVersion) throws EngineVersionException {
  if (minimumVersion == null) {
    return;
  }
  final Version mapCompatibleWithTripleaVersion =
      new Version(((Element) minimumVersion).getAttribute("minimumVersion"));
  if (mapCompatibleWithTripleaVersion.isGreaterThan(ClientContext.engineVersion(), true)) {
    throw new EngineVersionException(
        String.format("Current engine version: %s, is not compatible with version: %s, required by map: %s",
            ClientContext.engineVersion(),
            mapCompatibleWithTripleaVersion.toString(),
            data.getGameName()));
  }
}
```

I originally thought that the `isGreaterThan(Version, boolean)` call could be replaced with `isCompatible()`, but now I'm not positive.  The above check will be `false` if `mapCompatibleWithTripleaVersion` is less than the current engine version.  However, if `!isCompatible()` was used, the expression value would change to `true`.

There seem to be two distinct types of compatibility: engine-to-engine compatibility for communication (governed by `isCompatible()`) and map-to-engine compatibility for playing a specific map (governed by `isGreaterThan(Version, boolean)`).

RoiEXLab commented on src/main/java/games/strategy/util/Version.java:
```diff
@@ -131,42 +135,41 @@ public int hashCode() {
       }
     }
-    // if the only difference is m_micro, then ignore
     return 0;
```
Without the overload (and even with the ignoreMicro overload, but not that nicely and without being able to merge those 2 methods), we can replace those annoying if else statements with 4 nice lambda expressions using
```
Comparator.comparingInt(Version::getMajor)
  .thenComparingInt(Version::getMinor)
  .thenComparingInt(Version::getPoint)
  .thenComparingInt(Version::getMicro)
  .compare(this, other);
```
If we want to keep the overload the last lambda expression needs to be modified accordingly so the values passed to method are equal when ignoreMicro is true.

ssoloff replied:
Definitely and improvement for readability.  I'll try adopting that pattern.

ssoloff commented:
@RoiEXLab Please take a look at the latest commit and let me know if you think some of the code smells have been removed.

As I said above, there were two different engine compatibility checks.  Instead of trying to bury them inside `Version` itself, I extracted them to a new class (not sure if the name and/or package are the best choices).  Calls to the `isGreaterThan(Version, boolean)` and `isCompatible(Version)` methods were replaced with the appropriate newly-extracted method.

I added the `compareTo(Version, CompareOption...)` overload to support the newly-extracted compatibility methods.  Hopefully, the use of an enum instead of a boolean makes it a bit more readable.

I replaced calls to the `isGreaterThan(Version)` and `isLessThan(Version)` methods with direct calls to `compareTo(Version)` and the appropriate comparison with the return value.

RoiEXLab commented on src/main/java/games/strategy/util/Version.java:
```diff
@@ -131,58 +138,34 @@ public int hashCode() {
-    }
-    // if the only difference is m_micro, then ignore
-    return 0;
+    return compareTo(other, NO_COMPARE_OPTIONS);
```
I wondered why you're explicitly passing an empty array to this method, given that the argument is optional.

RoiEXLab replied:
(I figured out why, still seems weird to me)

RoiEXLab commented on src/main/java/games/strategy/engine/GameEngineUtils.java:
```diff
@@ -0,0 +1,43 @@
+    checkNotNull(engineVersion);
+    checkNotNull(otherEngineVersion);
+
+    return engineVersion.compareTo(otherEngineVersion, Version.CompareOption.IGNORE_MICRO) == 0;
```
I like the idea of extracting this kind of logic to a dedicated class, but in my opinion this doesn't go far enough:
I don't think such a single special case should be reason enough to handle them from within the Version class.
If we replaced this return statement with 3 Version#getMajor Minor and Point equals checks, we could drop all of the ignoreMicro logic and achieve the exact same result.

ssoloff replied:
I tend to agree with you, and, in fact, I briefly considered doing something similar to this before pushing the previous commit, but I felt that the amount of duplicate logic in the `isEngineCompatibleWithMap()` implementation would have been a smell.  However, that was before I introduced your suggested change to use `Comparator#comparingInt()` et. al.  Maybe that will clean it up a bit.  Let me give it another iteration to see if that works.

(Also, since we're talking about logic that doesn't belong in `Version` proper, I'm wondering if we should revisit how we handle the `dev` qualifier.  Now that I've been mucking with this class, it seems like something a bit too specific for this general-purpose type.  But let's save that for another PR. :smile:)

RoiEXLab commented:
And btw.: I'm fine with the package + class name

ssoloff commented:
@RoiEXLab I implemented your most-recent suggestion in the latest commit.  Let me know if you think it's not DRY enough.  I think it's just at the edge of acceptability, but I'm not sure if there's much more that can be done about it.

DanVanAtta commented on src/test/java/games/strategy/engine/GameEngineUtilsTest.java:
```diff
@@ -0,0 +1,37 @@
+public final class GameEngineUtilsTest {
+  @Test
+  public void testIsEngineCompatibleWithEngine() {
+    assertTrue(isEngineCompatibleWithEngine(new Version(1, 9), new Version(1, 9)));
```
It may be easier to add more test cases to be more exhaustive by using a lambda foreach loop as shown below:
```
Arrays.stream(
  new Version(1,9),
  new Version(1,9,0,1))
.forEach(compatibleValue)) {
 assertTrue(isEngineCompatibleWithEngine(new Version(1, 9), compatibleValue);
}
```
Then you could also verify yet more cases to go with that. For example it does seem like we should have some `assertTrue` cases with a major version of `2` to really be sure we have a contrast with the assertFalse cases.

ssoloff replied:
> Null comparsion checks look to be missing. If we fail to deserialize Version properly, it can be null, knowing the behavior of a null vs non-null would be useful.

As discussed elsewhere, without any clarification in the Javadocs, or the presence of a `@Nullable` annotation, callers should assume passing a `null` argument will result in an NPE.  Adding tests like `assertThrows(NullPointerException.class, obj.someMethod(null))` becomes quite a burden.  I worked on a codebase once where we did this, and it turned out that fully 50% of the unit tests were NPE checks.  A few years later, after we switched to a compiler that employed null analysis, we realized just how little value these tests were providing. :smile:

I would argue that a "deserialization failure" should be handled close to the point of failure instead of propagating `null` throughout the system and trying to come up with some rules for handling it.  In that case, raising an NPE would help us detect our "failure" to handle the failure.

> A secondary suggestion

I'll update the tests appropriately.

DanVanAtta replied:
Well said regarding null and its handling, I agree. As a side note, we have some work to do somewhere to fix the `Version` deserialization case.

@ssoloff  I'm curious, was your team also able to get rid of `Optionals` in the same way you removed the tests?

ssoloff replied:
> Well said regarding null and its handling, I agree.

You actually suggested we do that a couple of months ago.  I'm just riding your coat-tails. :smile:

>  I'm curious, was your team also able to get rid of Optionals in the same way you removed the tests?

That particular codebase was Java 6/7, so `Optional` wasn't in use.  When we had to represent "optional" data back then, we tended to use Null Objects instead of `null` whenever possible.

DanVanAtta commented on src/test/java/games/strategy/util/VersionTest.java:
```diff
@@ -1,106 +1,61 @@
-    assertFalse(new Version(1, 2, 3).isGreaterThan(null));
-    assertFalse(new Version(1, 2, 3).isLessThan(null));
+  public void testToStringFull() {
+    assertEquals("1_2_3_dev", new Version("1.2.3.dev").toStringFull('_'));
```
Any reason to not check the other `new Version` constructor variants as well in this test case? My concern is we may not be testing `toStringFull` as thoroughly as we could be otherwise.

ssoloff replied:
No reason other than that this was an existing test and had nothing to do with this PR. :smile:  It only appears as a change because it got moved around when I reorganized `VersionTest`.

DanVanAtta commented on src/test/java/games/strategy/util/VersionTest.java:
```diff
@@ -1,106 +1,61 @@
-    assertNotEquals(v2, v1);
-  }
+  public void testCompareTo() {
+    assertThrows(NullPointerException.class, () -> new Version(1, 0).compareTo(null));
```
This test I think assumes an implementation detail, that the constructors are unified. If we were to refactor so this constructor were unique in null handling, but the others were then newly buggy for whatever reason, this test case would not catch the regression. To fix that we could add the additional `new Version` constructor variants to this test case, eg:
```
Arrays.stream(
  new Version(0,0),
  new Version(0,0,0),
  new Version(0,0,0,0))
.forEach(version ->  assertThrows(NullPointerException.class, () -> version.compareTo(null)));
```

ssoloff commented on src/main/java/games/strategy/engine/GameEngineUtils.java:
```diff
@@ -0,0 +1,52 @@
+   *
+   * @return {@code true} if the engines are compatible; otherwise {@code false}.
+   */
+  public static boolean isEngineCompatibleWithEngine(final Version engineVersion, final Version otherEngineVersion) {
```
These methods were originally extracted because the definition of engine-engine compatibility and engine-map compatibility was buried in the `Version` class.  `Version` is used throughout the code to represent many different types of versions (e.g. engine, map, lobby protocol, etc.).  It wasn't clear why we were promoting functionality specific to engine versions in what should just be a simple value type representing a four-component version.

I agree that there is probably a leaky implementation here regarding the comparison.  If you go back to f4ca9dd, you'll see that it was still encapsulated within `Version` at the expense of providing an option to ignore the micro version during comparison.  I think the argument for getting rid of that option was that it was yet another implementation detail that had to do with some specific use cases (e.g. engine-engine compatibility) and had nothing to do with the general `Version` functionality.

Here's another way we could address these competing forces: provide an method on `Version` that returns the instance without the micro version:

```java
public Version withoutMicro() {
  return new Version(m_major, m_minor, m_point, 0);
}
```

Then the comparison logic within `Version` doesn't need to be duplicated outside that class and their need be no special overloads that accept options for ignoring certain fields.  For example:

```java
engineVersion.withoutMicro().compareTo(otherEngineVersion.withoutMicro()) == 0
```

Actually, with this approach, you can just re-use `equals()` in this situation (which we couldn't do before with any of the previous approaches):

```java
engineVersion.withoutMicro().equals(otherEngineVersion.withoutMicro())
```

DanVanAtta replied:
The suggestion here sounds good :+1:

ssoloff commented on src/main/java/games/strategy/engine/GameEngineUtils.java:
```diff
@@ -0,0 +1,52 @@
+   *
+   * @return {@code true} if the engine is compatible with the map; otherwise {@code false}.
+   */
+  public static boolean isEngineCompatibleWithMap(final Version engineVersion, final Version mapMinimumEngineVersion) {
```
Agreed about the readability.  However, I still question if we should have _engine_-specific version methods in the _general_ `Version` class.

What if, instead of the static methods in `GameEngineUtils`, we had a kind of throwaway wrapper class for game engine versions to host these methods?  Something like:

```java
Version version = // ...
Version otherVersion = // ...

GameEngineVersion.of(version).isCompatibleWithEngine(otherVersion)
GameEngineVersion.of(version).isCompatibleWithMap(otherVersion)
```

DanVanAtta replied:
I'd just as well attach those methods to `Version.java` , but I've no strong objection to attaching to a wrapper `GameEngineVersion`. The biggest key for me was readability and notably converting from static methods to instance.

Possibly a side note, but in case it is relevant: MapVersion is interesting and we may want to keep in mind the future plans. The current map version spec I don't think is that intuitive, nor does it allow for a good degree of backward compatible changes. I'd like for us having a parser strategy to replace the current map XML parser that would allow us to have multiple parsers that could function across multiple different file specs.

ssoloff commented on src/main/java/games/strategy/engine/data/GameParser.java:
```diff
@@ -187,13 +188,12 @@ private void parseMinimumEngineVersionNumber(final Node minimumVersion) throws E
-    if (mapCompatibleWithTripleaVersion.isGreaterThan(ClientContext.engineVersion(), true)) {
+    final Version mapMinimumEngineVersion = new Version(((Element) minimumVersion).getAttribute("minimumVersion"));
+    if (!GameEngineUtils.isEngineCompatibleWithMap(ClientContext.engineVersion(), mapMinimumEngineVersion)) {
       throw new EngineVersionException(
```
Interesting point, but I'm still concerned about embedding domain-specific knowledge (i.e. what makes an engine compatible with a map) in `Version`.

DanVanAtta replied:
I'm wondering out loud a bit how the proposed `GameEngineVersion` and existing `Version.java` would interact. `Version.java` is essentially a value object. As such parsing methods and core equality and hashcode would belong. I suspect everything else is more in the engine domain and should then be moved.

From my perspective, we have two domains, engine and map data. I think we see an example in TripleA where the two domains use the same Version model when it should be two models instead.

That implies to me there should be a `GameEngineVersion` and a `MapEngineVersion` object, or two `Version` objects under different package subtrees (engine and map). Without taking that step directly in one go, I would personally rename `Version` to `GameEngineVersion`, or leave it named `Version`, aggregate everything there and then work to have an independent `MapEngineVersion` object. That object could then be pretty nice to provide factory methods for returning version specific parser objects.

Bit of an open-ended comment here, I think we can go a number of ways.  I'm mostly concerned that we avoid the static utility class and get the readability benefit. Anything that gets us that I'm quite happy with, whether that be a `Value` + `GameEngineVersion` object combo, or a `MapVersion` + `GameEngineVersion` combo.

ssoloff replied:
:+1:

ssoloff commented on src/main/java/games/strategy/engine/framework/map/download/MapDownloadList.java:
```diff
@@ -32,7 +32,7 @@
       if (mapVersion.isPresent()) {
         installed.add(download);
-        if (download.getVersion() != null && download.getVersion().isGreaterThan(mapVersion.get())) {
+        if (download.getVersion() != null && download.getVersion().compareTo(mapVersion.get()) > 0) {
```
Agreed.  I will re-introduce the appropriate methods on `Version`.

ssoloff commented:
@DanVanAtta The latest commits address most of your concerns in order from least controversial to most controversial. :smile:

ssoloff commented on src/main/java/games/strategy/engine/data/GameParser.java:
```diff
@@ -187,13 +188,13 @@ private void parseMinimumEngineVersionNumber(final Node minimumVersion) throws E
-    if (mapCompatibleWithTripleaVersion.isGreaterThan(ClientContext.engineVersion(), true)) {
+    final Version mapMinimumEngineVersion = new Version(((Element) minimumVersion).getAttribute("minimumVersion"));
+    if (!GameEngineVersion.of(ClientContext.engineVersion())
+        .isCompatibleWithMapMinimumEngineVersion(mapMinimumEngineVersion)) {
```
I'm wondering what the history is on why we ignore the micro version when checking engine-map compatibility since it's a _minimum_ version check, not an _absolute_ version check.  Why do we not simply do a straight-up comparison:

```java
if (ClientContext.engineVersion().isLessThan(mapMinimumEngineVersion)) {
  throw // ...
}
```

For example, if a map author went to the trouble of stating the minimum engine version was 1.9.0.1 (possibly because a new feature was added in that release), why do we allow the map to run in engine version 1.9.0.0?  (In practice, this probably doesn't matter much, as all the maps I've seen only specify major and minor components for the minimum engine version.)

Just wondering if we can get rid of the custom engine-map compatibility check (currently in `GameEngineVersion#isCompatibleWithMapMinimumEngineVersion()`) and rely just on vanilla `Version` methods.

RoiEXLab replied:
Agreeing the current system doesn't make a lot of sense, but "fixing it" doesn't solve all problems. For example (AFAIK): If a map author increments the version number because of a change that requires a newer triplea version, it will still be downloaded because the game engine can't tell for what version the map was made + there's no alternative and triplea always downloads from the latest master anyways.
I'll create a dedicated issue for this...

DanVanAtta commented:
As-is this looks good, rebasing and merging.
